### PR TITLE
Use native file I/O in local and overlay drive

### PR DESCRIPTION
**Pull request by weirddan455**
# Description

This PR replaces the use of C library I/O (`fopen()` and friends) with native file I/O.  Windows uses Win32 API calls and everyone else uses POSIX.

I was also able to remove some hacky `ftell()` + `fseek()` nonsense that should not be needed since there's no library buffering.  The OS itself will still buffer but that should be fine.  All we care about is that a write followed by a read will return the new data.

I have some more work on the file I/O planned to better handle timestamps (which will depend on native file I/O) but I thought it would be best to put that in a separate PR and get this in first.

# Manual testing

Tested on Windows (both MSYS2/GCC and MSVC/Clang) and Linux.  Crystal Caves and Mortal Kombat saves work which had been problematic on Windows in the past.

Other games with save state tested:  Secret of Monkey Island 2, Quake, Doom, Space Quest 1

# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [x] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

### CODE REVIEW

File: include/dos_system.h
Comment (johnnovak) [inline include/dos_system.h:193]: `const NativeFileHandle`?

File: include/fs_utils.h
Comment (johnnovak) [inline include/fs_utils.h:43]: Maybe format it like this to make it look marginally nicer:

```cpp
#if defined(WIN32)
	#include <windows.h>

	using NativeFileHandle = HANDLE;
	// Cannot be constexpr due to Win32 macro
	#define InvalidNativeFileHandle INVALID_HANDLE_VALUE

#else // Linux, macOS
	using NativeFileHandle = int;
	constexpr NativeFileHandle InvalidNativeFileHandle = -1;
#endif
```

File: include/fs_utils.h
Comment (johnnovak) [inline include/fs_utils.h:64]: Just run the formatter on these. Then you'll get stuff like this:

```cpp
enum class NativeFileSeek { Set, Current, End };
```

File: include/fs_utils.h
Comment (johnnovak) [inline include/fs_utils.h:213]: Indeed 😎

File: include/fs_utils.h
Comment (johnnovak) [inline include/fs_utils.h:215]: I guess we could even just use `int` for all these byte sizes because of the FAT16 limits if it's easier overall.

File: include/fs_utils.h
Comment (johnnovak) [inline include/fs_utils.h:217]: Just `int offset`? Or `int64`?

File: include/fs_utils.h
Comment (johnnovak) [inline include/fs_utils.h:220]: `int64`?

File: include/fs_utils.h
Comment (johnnovak) [inline include/fs_utils.h:49]: `int64`?

File: src/misc/fs_utils_posix.cpp
Comment (johnnovak) [inline src/misc/fs_utils_posix.cpp:363]: Nice explanation 👍🏻 And yeah, this is the correct way to read a file on all platforms. Even Linux might change things up tomorrow.

File: src/misc/fs_utils_posix.cpp
Comment (johnnovak) [inline src/misc/fs_utils_posix.cpp:370]: `num_bytes_read`?

File: src/misc/fs_utils_posix.cpp
Comment (johnnovak) [inline src/misc/fs_utils_posix.cpp:364]: `num_bytes_requested`?

File: src/misc/fs_utils_posix.cpp
Comment (johnnovak) [inline src/misc/fs_utils_posix.cpp:380]: `num_bytes_requested`?

File: src/misc/fs_utils_posix.cpp
Comment (johnnovak) [inline src/misc/fs_utils_posix.cpp:386]: `num_bytes_written`?

File: src/misc/fs_utils_posix.cpp
Comment (johnnovak) [inline src/misc/fs_utils_posix.cpp:399]: Space after `switch`.

Comment (johnnovak) [reply src/misc/fs_utils_posix.cpp:399]: Well, just reformat the whole file; formatting is off.

File: include/fs_utils.h
Comment (johnnovak) [inline include/fs_utils.h:47]: `NativeFileIoResult` would be better.

File: include/fs_utils.h
Comment (johnnovak) [reply include/fs_utils.h:49]: `num_bytes` would be better.

File: include/fs_utils.h
Comment (johnnovak) [inline include/fs_utils.h:53]: `NativeSeek`?

**Comment by shermp** [comment]
Is there a case for making these file i/o wrappers more generic so that other parts of the code could use them later. I'm talking about seek, read, write etc.

Other code that could use this would be the disk image code, in which case large file support would definitely be appreciated.

File: src/misc/fs_utils_posix.cpp
Comment (weirddan455) [reply src/misc/fs_utils_posix.cpp:363]: Linux is pretty strict about not breaking userspace so I doubt this behavior will change.  I suppose someone could mount their DOSBox drive on a networked filesystem or something though and then I'm not sure what the behavior is but the loop should work everywhere.

File: include/fs_utils.h
Comment (weirddan455) [reply include/fs_utils.h:215]: I suppose I can go up to `int64_t`.  I would need to write some additional code to break up large (greater than 4GB) reads/writes on Windows into multiple calls but that'll be easy enough.

**Comment by weirddan455** [comment]
> Is there a case for making these file i/o wrappers more generic so that other parts of the code could use them later. I'm talking about seek, read, write etc.
> 
> Other code that could use this would be the disk image code, in which case large file support would definitely be appreciated.

They're already fairly generic and should be fine to use elsewhere.  Some of the calls (especially create and open) may need additional flags depending on the use-case but I think we can cross that bridge when we come to it.

Maybe I should rename the functions to `create_native_file` `read_native_file`, etc.  I'll bump the size types up to `int64_t` as @johnnovak suggested as well.

File: include/fs_utils.h
Comment (FeralChild64) [reply include/fs_utils.h:213]: @johnnovak And the maximum file size for FAT32 is still 4GB.

**Review by FeralChild64 — COMMENTED** [review]
Looks OK. I have no remarks.

So far tried some Windows 3.1 software, Colonization game. and MkS_Vir virus scanner, didn't find any problems/

File: include/dos_system.h
Comment (weirddan455) [reply include/dos_system.h]: fixed

File: include/fs_utils.h
Comment (weirddan455) [reply include/fs_utils.h:45]: fixed

File: include/fs_utils.h
Comment (weirddan455) [reply include/fs_utils.h:59]: formatted

File: include/fs_utils.h
Comment (weirddan455) [reply include/fs_utils.h]: Removed this comment.  It can handle large files now.

File: include/fs_utils.h
Comment (weirddan455) [reply include/fs_utils.h]: Changed to `int64_t`

File: include/fs_utils.h
Comment (weirddan455) [reply include/fs_utils.h]: Changed to `int64_t`

File: include/fs_utils.h
Comment (weirddan455) [reply include/fs_utils.h]: Changed to `int64_t`

File: include/fs_utils.h
Comment (weirddan455) [reply include/fs_utils.h]: Changed to `int64_t` and renamed

File: src/misc/fs_utils_posix.cpp
Comment (weirddan455) [reply src/misc/fs_utils_posix.cpp]: renamed

File: src/misc/fs_utils_posix.cpp
Comment (weirddan455) [reply src/misc/fs_utils_posix.cpp]: renamed

File: src/misc/fs_utils_posix.cpp
Comment (weirddan455) [reply src/misc/fs_utils_posix.cpp]: renamed

File: src/misc/fs_utils_posix.cpp
Comment (weirddan455) [reply src/misc/fs_utils_posix.cpp]: renamed

File: src/misc/fs_utils_posix.cpp
Comment (weirddan455) [reply src/misc/fs_utils_posix.cpp]: formatted

File: include/fs_utils.h
Comment (weirddan455) [reply include/fs_utils.h]: renamed

File: include/fs_utils.h
Comment (weirddan455) [reply include/fs_utils.h]: renamed

**Comment by weirddan455** [comment]
Thanks for testing @FeralChild64 

I pushed changes addressing the comments above.  Mostly re-naming things and fixing formatting.  I also changed to `int64_t` everywhere in these native I/O functions.

They should be suitable for use where-ever else in the code they might be needed.  I renamed them as I said above to `open_native_file` and such (removing the `local_drive` prefix I had before).  I also made the fat attributes on `create_native_file` a `std::optional` as that one doesn't make sense if we want to create a normal file on the host.

File: src/dos/drive_local.cpp
Comment (johnnovak) [inline src/dos/drive_local.cpp:652]: Could you test this before merging?

**Review by johnnovak — APPROVED** [review]
This is solid step in the right direction @weirddan455 🎉 Looks very good now, left one last comment about testing Blackthorne if you have the capacity. Great work!

File: src/dos/drive_local.cpp
Comment (weirddan455) [reply src/dos/drive_local.cpp]: Tested.  It worked with this hack in and failed to launch with it removed.  I pushed a new commit with more accurate/less hacky behavior and tested this works on Linux and Windows.

I was side-eyeing that `reinterpret_cast` and it turned out to be incorrect for `SEEK_SET`.  The game seeks to 2777296400 (over 2GB) into the file.  Windows and POSIX both allow this.  It doesn't actually resize the file on seek but if you then write to the file, it will fill with zeros up to that size.

Anyway, we just need to interpret the value as unsigned if using `SEEK_SET`.  I put some references to MS-DOS Encyclopedia in the comments.

The error handling and set to 0 isn't needed by Blackthorne but may be needed by other DOS games doing weird things.  Apparently the DOS Seek interrupt never fails assuming you give it a valid handle and seek type.

**Comment by weirddan455** [comment]
I'll let you take a look at my last commit if you like @johnnovak addressing the Blackthrone issue.

**Comment by weirddan455** [comment]
Added a `LOG_WARNING` if the file seek fails.  I wasn't quite sure what to do in this case because apparently the real DOS interrupt never fails.  The most likely case is a game trying to seek before the beginning of the file so I'm just setting the file position to 0.

Blackthorn never hits this case.  Using unsigned integer lets it succeed.  I wanted this warning to be visible to end users rather than a debug log because it seems to be rare and I want users to put this in a bug report if my assumption proves to be incorrect.

File: src/dos/drive_local.cpp
Comment (johnnovak) [reply src/dos/drive_local.cpp]: Thanks for the detailed explanation @weirddan455. This seems to have been quite a rabbit hole, but I'm glad you got to the bottom of it 🎉

File: src/dos/drive_local.cpp
Comment (johnnovak) [inline src/dos/drive_local.cpp:683]: Very nice coments, @weirddan455, that's medal-worthy 🎖️

One minor thing, if you'd include the book page numbers next to the citation, that would be helpful. Entirely optional, though, if you're sick of this PR already 😆

**Review by johnnovak — APPROVED** [review]
Outstanding work @weirddan455, I'm glad you've managed to remove the Blackthorne hack as well 🚀 I agree with your last note about logging a warning, that's the best we can do.

File: src/dos/drive_local.cpp
Comment (weirddan455) [reply src/dos/drive_local.cpp:683]: Sure, added the page numbers.

File: src/dos/drive_local.cpp
Comment (weirddan455) [reply src/dos/drive_local.cpp]: It didn't actually take me too long.  Luckily it was an easily reproducible error and fail to launch on Blackthorne.  Then I just added a log statement and saw it was being cast to a large negative seek value, where it should have always been positive.

File: src/dos/drive_local.cpp
Comment (johnnovak) [reply src/dos/drive_local.cpp:683]: Noice 😎

**Pull request merged**
